[MRG] Replace verbose class with standard logging library

jklymak opened the pull request:
<!--Thank you so much for your PR! To help us review, fill out the form
to the best of your ability.  Please make use of the development guide at
https://matplotlib.org/devdocs/devel/index.html-->

<!--Provide a general summary of your changes in the title above, for
example "Raises ValueError on Non-Numeric Input to set_xlim".  Please avoid
non-descriptive titles such as "Addresses issue #8576".-->

<!--If you are able to do so, please do not create the
PR out of master, but out of a separate branch.  See
https://matplotlib.org/devel/gitwash/development_workflow.html for
instructions.-->

## PR Summary

This replaces Matplotlib's `Verbose` class and calls with the [standard `logging` library](https://docs.python.org/3/library/logging.html).   Hopefully this will lead to more debugging information and better integration with downstream packages.  

See #9302 and discussion therein.


## How to use:

### For the end-user:

The old command-line `--verbose-debug` and `--verbose-helpful` are still respected (and `--verbose-info` was even added).  These are likely to be deprecated but are not for now.

Better is for a script to load and configure the `logging` library *before* importing matplotlib:

At its most basic:
```python

import logging
logging.basicConfig(level=logging.DEBUG)
import matplotlib.pyplot as plt
```
in the header should do it.  A slightly more useful formatting will give line numbers:

```python
import logging
Format = '%(name)s:%(lineno)5d - %(levelname)s - %(message)s'
logging.basicConfig(level=logging.DEBUG,
                        format=Format)
import matplotlib.pyplot as plt
```

The resulting console output will be something like:

```
matplotlib.backends:   89 - INFO - backend MacOSX version unknown
matplotlib.font_manager: 1335 - INFO - findfont: Matching :family=sans-serif:style=normal:variant=normal:weight=normal:stretch=normal:size=10.0 to Bitstream Vera Sans ('/opt/X11/share/fonts/TTF/Vera.ttf') with score of 0.050000
matplotlib.text:  768 - INFO - x and y are not finite values for text string 'boo!'.  Not rendering text.
matplotlib.text:  768 - INFO - x and y are not finite values for text string 'boo!'.  Not rendering text.
```
which is then pretty easily grep-able to find debug info in particular modules.  

### For the module writer

At the top of the module, you need to `import logging`.  Then calls like

```python
log = logging.getLogger(__name__)
log.info('Here is some information')
log.debug('Here is some more detailed information')
```

will log to a logger named `matplotlib.yourmodulename`.  Theoretically, `getLogger` is not too expensive, so you need not save the logger for each class (and you don't need to instantiate it in any way other than a call to `getLogger` somewhere).

I've used the name `log` throughout the code, which maybe is a problematic choice.  One can use any name you want if this clashes in your module's namespace.    

#### Which logging level to use?

There are five logging levels.  `logging.critical()` and `logging.error()` are really only there for errors that will end the use of the library but not kill the interpreter.  `WARNING` overlaps with the `warning` library.  The [logging tutorial](https://docs.python.org/3/howto/logging.html#logging-basic-tutorial) has suggested differentiation between `logging.warning()` and `warning.warn()`.  

By default, `logging` displays all log messages at levels higher than `logging.WARNING` to `sys.stderr`.  

`logging.info()` is not displayed by default.  Its for information that the user may want to know if the program behaves oddly.  For instance, if an object isn't drawn because its position is `NaN`, that can usually be ignored, but a mystified user could set `logging.basicConfig(level=logging.INFO)` and if the module writer is kind get an error message that says why.  

`logging.debug()` is the least likely to be displayed, and hence can be the most verbose.

### Fate of --verbose and the rcParams...

The `--verbose-debug` and `--verbose-helpful` command line options are still handled and set 
`logging.basicConfig(level=logging.DEBUG)` and ``logging.basicConfig(level=logging.INFO)` respectively.  

The entries in `matplotlibrc` are still respected by the same logic (`verbose.level` and `verbose.fileo`). However, they are on the `deprecated` list and those entries have been removed from the `matplotlibrc.template`.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant 
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or the
recommended next step seems overly demanding , or if you would like help in
addressing a reviewer's comments.  And please ping us if you've been waiting
too long to hear back on your PR.-->

anntzer commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1442,7 +1445,7 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
     command = [str("latex"), "-interaction=nonstopmode",
                '"%s"' % latexfile]
-    verbose.report(command, 'debug')
+    log.debug(command)
```
Please use `log.debug('%s', command)` (throughout) so as not to crash if command has the bad idea of including a `%`.  See https://github.com/sphinx-doc/sphinx/issues/4070 for a similar issue (except that they mangled the fix :-/).

jklymak replied:
OK, I can see this has an advantage.  OTOH, there is lots of this kind of formatting that was being passed to `verbose`.  I don't mind fixing the easy cases, but there are also cases like:

```python
log.info('figure size (inches) has been adjusted '
                               'from %s x %s to %s x %s' % (wo, ho, w, h))
```

I tested this, and its not as simple as:

```python 
log.info('figure size (inches) has been adjusted '
                               'from %s x %s to %s x %s', (wo, ho, w, h))
```
which fails.   

More importantly, I'm concerned that by redoing all these I'll screw something up.  These aren't the easiest parts of the code to debug since the tests don't get run with the logging turned on.  

For now, I'll fix the easy one-`%s` cases, but not the more complicated formatting.

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -1036,22 +1136,22 @@ def rc_params_from_file(fname, fail_on_error=False, use_default_template=True):
+    log = logging.getLogger(__name__)
     if "".join(config['text.latex.preamble']):
-        verbose.report("""
+        log.info("""
```
The standard way is to do
```
log.info("format_string %s", args)
```
instead of
```
log.info("format string %s" % args)
```
Note that the latter form will raise an exception if args has the bad idea of including a `%` (see comment below for more details).

efiring replied:
Additional point of information: the reason the logging module uses a format string followed by arguments is that this means the string interpolation is done only if the logging level is such that the result will actually be used.  In other words, it is a performance optimization.

jklymak commented:
Asking for functionality reviews before I write up docs as they will need to be changed in a number of places if this is the way we want to go forward.

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+                  output:
+                  `import logging`
+                  `logger=logging.getLogger('matplotlib')`
+                  `logger.set_level(logging.INFO)`
```
import logging
logging.getLogger('matplotlib').setLevel(logging.INFO)

yes it's camelcase
and if you want to keep it over three lines, spaces around =

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+    """
+    levelmap = {'silent': logging.WARNING, 'helpful': logging.INFO,
+                'debug': logging.DEBUG, 'debug-annoying': logging.DEBUG,
+                'info': logging.INFO, 'warning': logging.warning}
```
logging.WARNING

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+                    fileo = open(file_str, 'w')
+                    # if this fails, we will just write to stdout
+                except IOError:
+                    warning.warn('could not open log file "{0}"'
```
warnings

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+                    # if this fails, we will just write to stdout
+                except IOError:
+                    warning.warn('could not open log file "{0}"'
+                                'for writing.\nCheck your '
```
I tend not to prefer embedding newlines into this kind of messages because you never know what the width of the console is going to be anyways.  Not a big deal though.

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+
+    for arg in sys.argv[1:]:
+        # cast to str because we are using unicode_literals,
+        # and argv is always str
```
I don't think you need the cast?  On Py2 the str will just happily check for startswith against unicode; on Py3 everything is str.
If you do need the cast I'd just write `for arg in map(str, sys.argv[1:])`

tacaswell replied:
This might a 2.6 hold over?

This works as expected on as far back as 2.7.3 (the earliest version conda had easily available).

tacaswell replied:
Works with 2.6.8 as well, I do wonder how far back this goes...

jklymak replied:
I'm going to leave this in as I just copied from old code and don't understand the cast.  Feel free to tell me the proper way to do it.

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,78 @@ def _is_writable_dir(p):
+                  'info', 'warning')
+
+        if arg.startswith(str('--verbose-')):
+            level_str = arg[10:]
```
10 => len()

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -311,7 +384,32 @@ def ge(self, level):
+    output through logger
+
+    if always is True, the report will occur on every function
+    call; otherwise only on the first time the function is called
```
I don't think we ever use always=True so we may as well drop that

jklymak replied:
The whole class is deprecated, so...

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -760,7 +858,8 @@ def gen_candidates():
 
-_obsolete_set = {'text.dvipnghack', 'legend.isaxes'}
+_obsolete_set = {'text.dvipnghack', 'legend.isaxes', 'verbose.level',
+                 'verbose.fileo'}
```
Why did you deprecated them?  I the log level should stay configurable from the mplrc?

anntzer commented on lib/matplotlib/backends/backend_gtk3.py:
```diff
@@ -3,7 +3,7 @@
 import six
 
-import os, sys
+import os, sys, loging
```
tyop

anntzer commented:
There's bunch of minor edits here and there (haven't listed them all) and I think the logger level should stay configurable from the matplotlibrc, but the general approach looks good to me.

anntzer commented:
> WARNING overlaps with the warning library. The logging tutorial has suggested differentiation between logging.warning() and warning.warn().

The main point that must be paid attention to is that by default, a given warning.warn (same content and emitted from the same line of code) will only be shown once (it is then stored in a cache and later calls will be suppressed), whereas a logger.warning will be shown every time.
logger.warning is also more easy to configure (together with all other logging settings) so I would prefer defaulting to using logger.warning, except for things that we explicitly want to make harder to suppress...

jklymak commented:
@anntzer  Thanks so much for the comments.  I'll try and address (though quite a few were just copied over, rather than new).  Your larger points:

More than happy to walk back the changes and deprecation warnings on the matplotlibrc.  I personally think it could cause more confusion than not, but maybe thats also how I've implemented it.  

Right now, it takes the most verbose of a) an already instantiated `matplotlib` logger (i.e by the user in their code), b) the command-line option and c) `matplotlibrc` file.  I could easily modify that if we aren't planning to deprecate the matplotlibrc pathway.  If we did that, I'd suggest that whatever is set in the matplotlibrc is *ignored* if either the logger has already been instantiated  of `--verbose-LEVEL` is used.  I'd probably keep the behaviour of choosing the most verbose between the command line and programatic invocation since they are both being actively sought by the user.  

I think there is also a problem calling all these things `verbose` when they are now explicitly calling `logging`.  But maybe thats a documentation issue.  

WRT guidance about what level of verbosity folks should use, and choices between `warnings.warn` and `logging.warning`, I'm *way* too inexperienced to make that call.  I wrote what I wrote above to start discussion, but it should really be the core developers who make that sort of policy.  (I'm also not volunteering to go through the code and decide what should get `logging.warning` versus `warnings.warn`!  I imagine thats a fix-as-we-go sort of thing, and again, I wouldn't presume to know what the right behaviour is)

anntzer commented:
I think we can leave the calls to warnings.warn as they are for now and change them to logger.warning in a piecemeal fashion whereever appropriate.
I think properly documenting that verbose is now controlling the mplrc is the easiest way to go.
I think the order of priority should be from low to high: matplotlibrc, --verbose-foo, explicitly configured logger.  The idea being that --verbose-foo is easier to set on an interactive basis than matplotlibrc, but if one already bothered setting up an explicitly configured logger, then adding the code to configure the level yourself should not be a big deal...

efiring commented:
I will try to give this a close look ASAP.  Thank you for taking this on; it's definitely the direction we need to go.

jklymak commented:
@efiring  That'd be great.  Did you want me to factor in @anntzer review first?

efiring commented:
Yes, please go ahead with @anntzer's suggestions.

tacaswell commented on lib/matplotlib/animation.py:
```diff
@@ -309,18 +310,17 @@ def frame_size(self):
             w, h = self.fig.get_size_inches()
-        verbose.report('frame size in pixels is %s x %s' % self.frame_size,
-                       level='debug')
+        log.info('frame size in pixels is %s x %s' % self.frame_size)
```
Did you mean to change the lever on this one (makes sense to match the one just above)

tacaswell commented:
Without minimizing the amount of work done, this is less invasive change than I expected!

I definitely think that this is the right way to go.  On a quick read, the mapping between the verbose levels and the logging level makes sense.

Agree with @anntzer to not touch `warnings.warn` for now.  Most of them are deprecation, not run time, warnings and we want to make them hard to suppress.

A while ago I also wrote up notes on loggers (https://github.com/scikit-beam/scikit-beam/blob/dd8625baa22c886601d51ff0c6fea0aee01c86cb/doc/resource/dev_guide/logging.rst ) which may or may not be a useful reference.

jklymak commented:
@anntzer  Above you proposed to prioritze:

  1. code-instantiation of logger
  2. command-line `--verbose-LEVEL` (but deprecate?)
  3. `maplotlibrc` `verbose.level` and `verbose.fileo`.

The problem is I don't think there is anyway from within `matplotlib.__init__` to know that the user has explictly done `logger = logging.getLogger('matplotlib').logger.set_level(logging.WARNING)` (you *can* check for other levels) because it is the default.  

If its true that we can't detect that the user has specified this in their code, then that becomes very confusing if we say that the coded logger will take priority *unless* it is set to `WARNING` and then we will look at the command line or matplotlibrc.

I am personally against this being set at all in the `matplotlibrc`.  How the library interacts with the interpreter doesn't seem like a role for the configuration file.  

My proposal would be:

  1. the level is the default `logging.WARNING`.
  2. the greater of the command-line argument and the explicitly coded argument.  
  3. the `matplotlibrc` is deprecated. 

If we want to keep the `matplotlibrc`, then I'd suggest the greater of the three pathways.  

This is probably just my bias, as I think I've dug out my `matplotlibrc` maybe twice in the 5 years I've used matplotlib ;-)

jklymak commented:
@tacaswell   Not minimizing the ammount of work at all.  It *was* pretty minimal.  Which I think is somewhat unfortunate, because it means people are avoiding putting debugging into their code.  Not sure if this will change that, and maybe there are good reasons not to spam the logger if you are doing something that requires high performance.

anntzer commented:
> @anntzer Above you proposed to prioritze:
> 
> code-instantiation of logger
> command-line --verbose-LEVEL (but deprecate?)
> maplotlibrc verbose.level and verbose.fileo.
> The problem is I don't think there is anyway from within matplotlib.__init__ to know that the user has explictly done logger = logging.getLogger('matplotlib').logger.set_level(logging.WARNING) (you can check for other levels) because it is the default.
> 
> If its true that we can't detect that the user has specified this in their code, then that becomes very confusing if we say that the coded logger will take priority unless it is set to WARNING and then we will look at the command line or matplotlibrc.

`"matplotlib" in logging.Logger.manager.loggerDict` (strictly speaking it only checks whether the logger already exists, but that seems close enough -- why would the user instantiate a logger otherwise?).

> I am personally against this being set at all in the matplotlibrc. How the library interacts with the interpreter doesn't seem like a role for the configuration file.
> 
> My proposal would be:
> 
> the level is the default logging.WARNING.
> the greater of the command-line argument and the explicitly coded argument.
> the matplotlibrc is deprecated.
> If we want to keep the matplotlibrc, then I'd suggest the greater of the three pathways.

I think I'm +0.5 on *keeping* matplotlibrc support, but am happy to hear input from others.

jklymak commented:
Awesome!  I’m fine with instantiation trumps all if that is the route to go.  If anyone else has an opinion let me know.  If not I’ll implement @anntzers preferred method.

jklymak commented:
Meh.  If we set `logging.basicConfig(level=logging.WARNING)` (which is the easiest way to invoke logging) then `logging.Logger.manager.loggerDict` stays empty, but the root logger is changed.

jklymak commented:
pinging @anntzer for advice as per the above comment.  I think its hard to tell if the `root` logger (as opposed to the `matpltolib`-specific logger) is configured so that makes it hard to prioritize it.  Happy to handle the prioritizing of debug level how the core devs would like....

anntzer commented:
I thought about the issue again with a fresh mind...

I now agree with @jklymak (though for different reasons) that configuration in matplotlibrc is not necessary.  Basically, if the user already bothered calling `logging.basicConfig()` themselves, then they should be able to set the level on the 'matplotlib' logger just as well.  In that context, we should just not bother setting any level on the 'matplotlib' logger.

As for `--verbose-foo`, if it is set at the command line then the implicit consequence is that matplotlib is already taking over (part of) the whole program's argv (which is the reason why the feature should be deprecated), and in that case it is only a small extra step to also instantiate ourselves a root handler at the correct level.

To be honest I'm not even convinced that there's much of a benefit in keeping the matplotlibrc entries / --verbose-foo flags for a deprecation period.  Usually the point of a deprecation period is to give users time to update their code instead of breaking it, but here it's not as if removing them immediately would "break" anything (... unless someone actually relies on the logs...).

jklymak commented:
@anntzer That would be great, and much simpler.

The only issue I can see is that it may be easier to tell a naive user to put a commandline option rather than muck with imports etc in their program. But I don't think Matplotlib has that many naive users...  I've also been reading the mailing list and looking at github entries for a couple of years now, and I don't recall too many instances of devs asking for the verbose output.  

I'll definitely update the `debugging matplotlib` section of the relevant manual if this is the way to go.  I'll wait a few days for more dev input.  ping @efiring who said he was interested in taking a look...

anntzer commented:
Actually (on a parallel note) on Py3 logs at a level >= WARNING get printed by default (https://docs.python.org/3/library/logging.html#logging.lastResort).  I think that just as good, but wanted to point this out for the discussion.

I think telling users to add `import logging; logging.basicConfig(level=logging.DEBUG)` is not too much harder than telling them to use --verbose-foo (especially as the latter will NOT work if they are also doing their own argument parsing).

anntzer commented on doc/faq/troubleshooting_faq.rst:
```diff
@@ -114,11 +114,39 @@ provide the following information in your e-mail to the `mailing list
+
+    import matplotlib.pyplot as plt
+
+  (the ``Format`` just gives better looking output and is optional).
```
For the purpose of troubleshooting, I would favor the much less complicated call without setting an explicit format (it's good enough for a bug report), and also not require to run a specific file.  Something like
```
python -c "from logging import *; basicConfig(level=DEBUG); from pylab import *; plot(); show()"
```
which should be good enough to debug many "startup" problems.

For debugging a script, I would likewise document `import logging; logging.basicConfig(level=logging.DEBUG)`.

anntzer commented on doc/users/next_whats_new/2017_10_10_logging.rst:
```diff
@@ -0,0 +1,24 @@
+------------------------
+
+Matplotlib has in the past (sporadically) used an internal
+verbose-output reporter.  This version converted those callse to using the
```
callse (typo)

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,77 @@ def _is_writable_dir(p):
+                "This functionality is now provided by the standard "
+                "python logging library.  To get more (or less) logging "
+                " output:\n"
+                "`import logging`\n "
```
probably no backquotes, or only around the whole expression.  no dot at the end of the next line

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +217,77 @@ def _is_writable_dir(p):
+                except IOError:
+                    warnings.warn('could not open log file "{0}"'
+                                'for writing.  Check your '
+                                'matplotlibrc'.format(file_str))
```
indent this consistently?

anntzer commented on lib/matplotlib/animation.py:
```diff
@@ -564,26 +562,25 @@ def finish(self):
             msg = ('Error creating movie, return code: ' +
                    str(self._proc.returncode) +
-                   ' Try setting mpl.verbose.set_level("helpful")')
+                   ' Try setting python logging level to INFO')
```
I think the logs about should just be moved to warning level?...

jklymak replied:
`helpful` was equivalent to `INFO` so thats what I used.  level is already WARNING by deafult if logging is turned on.

anntzer replied:
my point is that there isn't much of a point to have a warning-level message that says "this message is useless but you can replace it by a useful message by running with a different log-level setting" -- you may as well directly give the answer!

jklymak replied:
I see now.  Yes, I made those `log.error` and they should preceed the RuntimeError message.

anntzer commented on lib/matplotlib/animation.py:
```diff
@@ -645,7 +642,9 @@ def _args(self):
-        if not verbose.ge('debug'):
+        log = logging.getLogger(__name__)
+
+        if not (log.getEffectiveLevel() >= logging.DEBUG):
```
https://docs.python.org/3/library/logging.html#logging.Logger.isEnabledFor

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -118,12 +118,14 @@
 import sys
 import tempfile
 import warnings
+import logging
```
try to keep imports in alphabetical order

anntzer commented on lib/matplotlib/backends/backend_gtk3.py:
```diff
@@ -366,7 +366,8 @@ def __init__(self, canvas, num):
             # better way is - JDH
-            verbose.report('Could not load matplotlib icon: %s' % sys.exc_info()[1])
+            log = logging.getLogger(__name__)
+            log.info('Could not load matplotlib icon: %s', sys.exc_info()[1])
```
you can just do `log.info("Could not load matplotlib icon.", exc_info=True)` (magic! see https://docs.python.org/3/library/logging.html#logging.Logger.debug)

anntzer commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -658,9 +659,8 @@ def fontName(self, fontprop):
-                'Assigning font %s = %r' % (Fx, filename),
-                'debug')
+            logger = logging.getLogger(__name__)
+            logger.debug('Assigning font %s = %r' % (Fx, filename))
```
fmt string

anntzer commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1453,11 +1456,11 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
              '\n\n' % (latexfile,
                        exc.output.decode("utf-8"))))
-    verbose.report(report, 'debug')
+    log.debug(report)
```
fmt string throughout

anntzer commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1442,7 +1445,7 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
     command = [str("latex"), "-interaction=nonstopmode",
                '"%s"' % latexfile]
-    verbose.report(command, 'debug')
+    log.debug('%s', command)
```
we don't have to do it in this PR but I think we should just have our own wrapper around subprocess which does the logging every time we shell out

anntzer commented on lib/matplotlib/backends/backend_tkagg.py:
```diff
@@ -1060,7 +1060,8 @@ def new_figure_manager_given_figure(num, figure):
             # log the failure (due e.g. to Tk version), but carry on
-            verbose.report('Could not load matplotlib icon: %s' % exc)
+            log = logging.getLogger(__name__)
+            log.info('Could not load matplotlib icon: %s', exc)
```
see above re: exc_info

anntzer commented on lib/matplotlib/dviread.py:
```diff
@@ -859,7 +862,8 @@ def __getitem__(self, texname):
                                 break_long_words=False)
-            matplotlib.verbose.report(msg, 'helpful')
+            log = logging.getLogger(__name__)
+            log.info(msg)
```
I would not bother with the call to textwrap, and do the formatting here.

jklymak replied:
Sent a gitter, but I have no idea what this is supposed to do and am not comfortable re-doing the formatting.  I think we could have another PR that cleans up the log messages, but am concerned about doing too many formatting changes unless they are very obvious.

anntzer replied:
It's totally fine.  I think this PR is very close to being mergeable (with the rest to be fixed in a piecemeal fashion later).

jklymak replied:
Phew!  I wasn't looking froward to getting the font library messages all sorted...

anntzer commented on lib/matplotlib/blocking_input.py:
```diff
@@ -25,11 +25,10 @@
 from __future__ import (absolute_import, division, print_function,
                         unicode_literals)
 
+import logging
```
not to be a pain but https://github.com/matplotlib/matplotlib/pull/7454#discussion_r87736843 (yes that should probably be documented in the style guide)

jklymak replied:
The only pain is that I don't know what you mean here ;-)  should logging come after six?  You just told me to make everything alphabetical!  ;-)

anntzer replied:
The standard order should be

future imports + six
(empty line)
standard library
(empty line)
third party
(empty line)
matplotlib other modules

(~ https://www.python.org/dev/peps/pep-0008/#imports)

and try to keep things ordered in each section.

jklymak replied:
OK, I didn't clean up all the other imports that don't follow this, but I tried to fix `logging` to follow this...

jklymak commented:
Docs that changed:  https://2695-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/faq/troubleshooting_faq.html

https://2695-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/devel/contributing.html

anntzer commented:
You can avoid mishighlighting the output in the docs at the end of https://2695-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/devel/contributing.html#using-logging-for-debug-messages by using `.. code:: none` (http://docutils.sourceforge.net/docs/ref/rst/directives.html#code).

jklymak commented:
`.. code:: none` gives: `/Users/jklymak/matplotlib/doc/devel/contributing.rst:475:Cannot analyze code. No Pygments lexer found for "none".`

anntzer commented:
at least `code-block:: none` should work (it's explicitly documented at http://www.sphinx-doc.org/en/stable/markup/code.html#directive-code-block)

jklymak commented:
@efiring I'm open to changing that strategy.  @tacaswell  has the same comments.  

I'm not claiming to know the best way to do this *at all*.  I just did what made sense to me from a few quick googles on the web.  If you'd like to have a module-level instantiation, its fine with me.  I'm not even sure what other libraries that use logging do.  I'm not even sure what libraries use logging!

I don't think that the import-time is an issue!

The only disadvantage is that the logger gets set at module import time presumably, and that can be before the user configures the logging.  By just `getLogger` on an as-needed basis the user can set up logging at any time in the preamble, but before the methods in the class get called, so its less fragile.  

Second, is that developers don't need to keep track of whether or not a module has set up logging.  Thats admittedly a weaker argument because they have to at least `import logging` if it hasn't been imported yet.  But a "new" dev who wants to add logging to an existing module has to `import logging` *and* add a new logging object to the class, or a module global.

efiring commented:
Regarding the question of where to call `getLogger`:
I am leaning strongly toward the "make a logger at the top of the module" approach.  For the new developer this is no more of a burden than other development conventions, such as our standard `__future__` lines (which will go away soon--hooray!).  It is balanced by being able to use `_log.warning` etc. anywhere thereafter without checking to see whether a logger already exists in the local namespace, and without having to write two lines instead of one.  It makes using the logger more similar to using print statements or `warnings.warn`.
Regarding the question of who configures what, when, and where, I don't think it actually reduces flexibility or imposes additional constraints.  The `matplotlib` logger is already being generated at matplotlib import time, and having the module loggers (children of `matplotlib`) created at the module import time just means that they are known to exist when anything in a module is used.  They are still, by default, passing everything up the chain.  At any time, any logger, including but not limited to the root logger, can be configured or reconfigured, and that change will take effect right away.  In other words, I don't see the fragility you are referring to--but I might be missing something.

Sphinx is an example of a library/application using the logging module, and it uses this approach.  It uses the name `logger`.  PIL is another example.

jklymak commented:
OK, so at the top of the module you are proposing:

```python

import blahblah
import logging
import zaza

import matplotlib.stuff

logger = logging.getLogger(__name__)

def mylib():
    logger.debug('Hi There!')

```

Thanks I'll look at PIL and sphinx too....

jklymak commented:
@tacaswell @anntzer @QuLogic @dopplershift @WeatherGod @dstansby @choldgraf  @afvincent (Sorry if I misssed someone considered "core" or "active")  

Is the above OK before I go and fix it all?  FWIW I checked sphinx and PIL and that is exactly what they do - under the import blocks but before any classes/function they have a line `logger = logging.getLogger(__name__)` and then just log with `logger.info('Blah')` etc...  

Easy fix, but it's a bit of fiddling that I'd rather only do one more time!

choldgraf commented:
as a general rule, I am  +1 to using well-supported tools that are out there in the OS world, rather than writing our own code

anntzer commented:
I would make it private (`_logger` or `_log`), as suggested in https://github.com/matplotlib/matplotlib/pull/9313#pullrequestreview-71719639.  Otherwise I agree with the idea.

jklymak commented:
@anntzer I don't have a problem with `_logger`, except that the libraries above just use `logger`. It really doesn't matter to me, so long as people use it ;-)

WeatherGod commented:
Be sure to do a `logger.addHandler(logging.NullHandler())` right after the
`getLogger`. It'll help prevent warning messages to people using matplotlib
without setting any logger at all.

On Fri, Oct 27, 2017 at 12:11 PM, Jody Klymak <notifications@github.com>
wrote:

> @anntzer <https://github.com/anntzer> I don't have a problem with _logger,
> except that the libraries above just use logger. It really doesn't matter
> to me, so long as people use it ;-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/9313#issuecomment-340015769>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-BIiknkRHen8XiVuStaK5epZLh0Lks5swgCngaJpZM4Pxlg2>
> .
>

jklymak commented:
@WeatherGod  Well that’s a discussion item I guess.  I’d argue warnings should be warnings and should go to `stderr`.

WeatherGod commented:
I think you misunderstand. Here is documentation about what I mean:
https://docs.python.org/2/howto/logging.html#configuring-logging-for-a-library

On Fri, Oct 27, 2017 at 12:28 PM, Jody Klymak <notifications@github.com>
wrote:

> @WeatherGod <https://github.com/weathergod> Well that’s a discussion item
> I guess. I’d argue warnings should be warnings and should go to stderr.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/9313#issuecomment-340019755>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-JXpNaMWgBJaKEaXIgVA4EfPnBukks5swgSPgaJpZM4Pxlg2>
> .
>

anntzer commented:
If we go this route (... to be discussed), this can (should) be attached to the toplevel logger (so just once) rather than to each instance.

dopplershift commented:
@WeatherGod Can you clarify? Because in the linked section it says:

> A do-nothing handler is included in the logging package: NullHandler (since Python 3.1). An instance of this handler could be added to the top-level logger of the logging namespace used by the library (if you want to prevent your library’s logged events being output to sys.stderr in the absence of logging configuration)

This would seem to be the very opposite of what @jklymak is arguing for, and I agree with him.

WeatherGod commented:
I use NullHandlers all the time for my packages at work because often, my users are at the REPL and aren't concerned about logging configurations. The logging is mostly for my scripts that uses the packages and which does configure the logging. Warning messages about logging that my users didn't even know existed would only serve to confuse them, thinking that they are doing something wrong.

matplotlib is often used at a REPL of some sort, and the logging isn't really all that useful for end-users (although extremely useful for the devs). The lowest common denominator, which are end-users at the REPL, don't need to see warning messages about a logging system that is not intended for them.

And, yes, we can do a single NullHandler at the top-level module. I had gotten used to putting it with all getLogger calls for other devs who were unaware of python's logging tools and its ins-and-outs. They just copy-n-paste to their own projects and it "works" as expected.

WeatherGod commented:
Reading that documentation, I think there is a misunderstanding of what NullHandler does. In python2.7, if no handler is found at all, nothing is emitted except a message saying that no handler was found. Now, maybe that is different in py3k, but in py27, all NullHandler does is suppresses that annoying message.

WeatherGod commented:
Yup.... looks like a behavior difference between py27 and py3k. Ok, it makes sense to have NullHandler added just for py27, then?

jklymak commented:
Hmm, the 2.7 docs say the same thing, however...

>A do-nothing handler is included in the logging package: NullHandler (since Python 2.7). An instance of this handler could be added to the top-level logger of the logging namespace used by the library (if you want to prevent an error message being output to sys.stderr in the absence of logging configuration). 

I just installed a fresh py2.7 environement with this branch as the matplotlib and ran the following with no anoying messages about the handler.  So, am I not testing it properly?

```python
import matplotlib.pyplot as plt

fig = plt.figure()
plt.plot([5,6])
plt.show()
```

WeatherGod commented:
I think with the way you have it set up, a StreamHandler() is always added to the logger?

WeatherGod commented:
```
[broot@rd22 isd_loader]$ python
Python 2.7.12 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:42:40) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import logging
>>> log = logging.getLogger(__name__)
>>> log.info("foo")
>>> log.error("foo")
No handlers could be found for logger "__main__"
```

WeatherGod commented:
The example you had would not trigger the message because the messages that were getting emitted were not hitting the logging level threshold, so it wasn't looking for a handler.

jklymak commented:
Only if called via `--verbose-helpful` or `--verbose-debug`.  

Otherwise all I ever do in the library is `import logging`, `log = logging.getLogger(__name__)` and `log.info('my message')`, `log.debug('my message')`, or in *very* few places `log.warning('my message')`.

jklymak commented:
Yes, OK, confirmed by putting a warning in `axes/_axes/plot()`:

`No handlers could be found for logger "matplotlib.axes._axes"`

efiring commented:
The default behavior differs for python 3:
```
(test) efiring@manini2:~/Downloads$ python
Python 3.5.2 |Anaconda custom (x86_64)| (default, Jul  2 2016, 17:52:12)
[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import logging
>>> log = logging.getLogger(__name__)
>>> log.error('foo')
foo
>>> log.handlers
[]
>>> rootlog = logging.getLogger()
>>> rootlog.handlers
[]
>>> rootlog
<logging.RootLogger object at 0x102b39198>
>>> rootlog.name
'root'
>>> log.setLevel(logging.DEBUG)
>>> log.info('informative')
>>> rootlog.setLevel(logging.DEBUG)
>>> log.info('informative')
>>> log.warn('warning')
warning
```
So it seems that in Python 3, if no handler has been supplied, output goes to stderr *without* any handler being in the list, and its level is logging.WARN.

This is a tricky business: we want to make mpl's use of logging simple and understandable, but we don't want to clobber or complicate the use of logging in code that uses mpl.  My inclination here is to target Python 3, which provides this fallback output mechanism for warnings and errors, and to *not* set any handlers by default.  I think the further configuration, including setting handlers, needs to be done in the code using mpl, not in mpl itself.  Otherwise the external code might need to explicitly remove unwanted mpl-supplied handlers.

jklymak commented:
At the top of `__init__.py` 

```python
_logger = logging.getLogger(__name__)
if _python27:
    _logger.addHandler(logging.NullHandler())
```

seems to work fine to suppress the warning message w/o changing the python 3.   Configured loggers still work fine.  So that is a way forward if we want to give py27 users a cleaner experience.  OTOH, py27 users won't see the same warnings (and above) that py3 users will see (in either case).

WeatherGod commented:
yes, @jklymak, I think that is exactly what we should have. The way this can be thought of is that for py27, nothing changes for the end-user if no configs were ever set. py3k users get the added benefit of warning messages to stderr for free.

jklymak commented:
While folks are perusing this: `_logger` or `logger`?  @anntzer already on record for leading underscore.

efiring commented:
pip also uses "logger" (along with sphinx and pillow), but beyond that, a little grepping in site-packages hasn't revealed much systematic use of the logging module along these lines!  There are some other instances of setting a NullHandler for 2.7, as above.

The advantage of the leading underscore is that it prevents the name from being pulled in by "from foo import *", if someone chooses to do that, contrary to most advice.

I still have a mild preference for `_log`, to keep things short as well as private.

jklymak commented:
OK< I think this is passing, and ready to go:

Changed to `_log` and  put `_log = logging.getLogger(__name__)` at the earliest possible location below the import blocks.  Also added the `NullHandler` for Python 2.7 in `__init__.py`.  

I think this is ready to go, but happy to fix other issues.

jklymak commented:
ping @efiring to re-review.  This one has many hooks, so it'd be nice to merge so conflicts don't get overwhelming....

efiring commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +222,76 @@ def _is_writable_dir(p):
+                " output:\n"
+                "import logging\n "
+                "logger = logging.getLogger('matplotlib')."
+                "logger.set_level(logging.INFO)")
```
Looks like you need to add newlines in this line and the one above.

anntzer replied:
Also remove the dot at the end of the previous line -- or just squeeze the whole thing into `logging.getLogger('matplotlib').set_level(logging.INFO)`
Also remove the space in `" output:"`.

TBH I would consider using a triple quoted string here:
```
# note the backslash in the first line!
_verbose_msg = """\
--verbose-LEVEL and the Verbose class etc.
import logging
logging.getLogger('matplotlib').set_level(logging.INFO)"""
```

efiring commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +222,76 @@ def _is_writable_dir(p):
+    # Check that current state of logger isn't already more verbose
+    # than the requested level.  If its more verbose, then leave more
+    # verbose.
+    if level_str in levelmap:
```
This check is actually performed in each case by upstream calling code, correct?

jklymak replied:
Yes...  Well, it is now that I remembered to remove the call to this in `rc_params_from_file`

efiring commented on lib/matplotlib/animation.py:
```diff
@@ -314,13 +317,11 @@ def _adjust_frame_size(self):
-                               'from %s x %s to %s x %s' % (wo, ho, w, h),
-                               level='helpful')
+                _log.info('figure size (inches) has been adjusted '
+                               'from %s x %s to %s x %s' % (wo, ho, w, h))
```
At some point these should be converted to use the multiple arguments syntax of logging rather than the original string substitution syntax.  Do you want to defer that?

anntzer replied:
There's also the issue of reindenting everything.
I think this can wait, it's easy enough to grep for `\b_log.` and fix this in a piecemeal fashion.

jklymak replied:
I just avoided doing this because my formatting skills are poor.  In this case can I just do:
```python
_log.info('figure size (inches) has been adjusted '
                               'from %s x %s to %s x %s', (wo, ho, w, h))
```

jklymak replied:
I'll note its not slowing down relative to before where these formatters were all being executed, so at least by not fixing these I'm not making things worse.

anntzer replied:
the correct fix would be
```
_log.info('figure size (inches) has been adjusted '
                               'from %s x %s to %s x %s', wo, ho, w, h)
```
(no inner parentheses)

efiring commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1431,9 +1434,10 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
-                verbose.report("You are using unicode and latex, but have "
-                               "not enabled the matplotlib 'text.latex.unicode' "
-                               "rcParam.", 'helpful')
+
```
Picky: delete blank line here.

efiring commented on lib/matplotlib/text.py:
```diff
@@ -761,9 +764,7 @@ def draw(self, renderer):
-                verbose.report("x and y are not finite values for text "
-                              "string '{}'.  Not rendering "
-                              "text.".format(self.get_text()), 'helpful')
+                _log.warning("posx and posy should be finite values")
```
I suspect this change in the message content has already been discussed and approved: correct?

jklymak replied:
Yes - umm, I forget where, but I pushed this change...

anntzer commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1453,11 +1457,11 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
     command = [str('dvips'), '-q', '-R0', '-o', os.path.basename(psfile),
                os.path.basename(dvifile)]
-    verbose.report(command, 'debug')
+    _log.debug(command)
```
I'm actually surprised that this even works...

jklymak replied:
Ummm....  It worked before?  I don't actually have a good way of checking if it works here.  Or can I just print an eps in debug mode?

anntzer replied:
What I mean is that I didn't know you could pass a list to log.info and it'll just call str() on it itself (which seems like a misfeature but whatever).

It may make sense to set the log level to debug (while piping it to devnull :-)) on one of the travis job, to check that logging actually works :-) but that can wait too IMO.

jklymak replied:
Yeah, these are all in `convert_psfrags`, and I don't even know how to get that fcn to call ;-)

QuLogic replied:
Why did you convert the one above to use `'%s'` and not this one?

anntzer commented on doc/devel/contributing.rst:
```diff
@@ -437,6 +437,73 @@ forced to use ``**kwargs``.  An example is
+Using logging for debug messages
+--------------------------------
+
+Matplotlib uses the standard python logging_ library to write verbose
```
We have an intersphinx mapping (https://github.com/matplotlib/matplotlib/blob/master/doc/conf.py#L95) to the stdlib so you can just write
```
`logging`
`print`
`warnings.warn`
```
etc. and they will link to the corresponding entries in the stdlib docs

anntzer commented on doc/devel/contributing.rst:
```diff
@@ -437,6 +437,73 @@ forced to use ``**kwargs``.  An example is
+Which logging level to use?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are five logging levels.  :func:`logging.critical()` and
```
the level is logging.CRITICAL, not logging.critical.

jklymak replied:
OK, but I want to refer to the functions.  Changed to `There are five levels at which you can emit messages.`

anntzer commented on doc/faq/troubleshooting_faq.rst:
```diff
@@ -114,11 +114,37 @@ provide the following information in your e-mail to the `mailing list
+
+    import matplotlib.pyplot as plt
+
+  Note that if you want to use logging_ in your own code, but do not
```
see above re: intersphinx

anntzer commented on .appveyor.yml:
```diff
@@ -144,7 +144,7 @@ after_test:
 
   # Build the wheel with the static libs
-  # Hide the output, the copied files really clutter the build log...
+  # Hide the output, the copied files really clutter the build_log...
```
leave this as it was?

anntzer commented on doc/devel/contributing.rst:
```diff
@@ -437,6 +437,71 @@ forced to use ``**kwargs``.  An example is
+There are five levels at which you can emit messages.
+`logging.critical` and `logging.error`
+are really only there for errors that will end the use of the library but
+not kill the interpreter.  :`logging.warning()` overlaps with the
```
spurious colon

anntzer commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +222,76 @@ def _is_writable_dir(p):
+_verbose_msg = """\
+Command line argument --verbose-LEVEL is deprecated.
+This functionality is now provided by the standard
+python logging library.  To get more (or less) logging output:"
```
spurious quote

anntzer commented on lib/matplotlib/animation.py:
```diff
@@ -650,7 +640,8 @@ def _args(self):
         # Logging is quieted because subprocess.PIPE has limited buffer size.
-        if not verbose.ge('debug'):
+
+        if not (_log.getEffectiveLevel() >= logging.DEBUG):
```
aka < DEBUG? :-)

jklymak replied:
I guess that works!

QuLogic commented on lib/matplotlib/animation.py:
```diff
@@ -933,7 +924,7 @@ def __init__(self, fps=30, codec=None, bitrate=None, extra_args=None,
         if self.default_mode not in ['loop', 'once', 'reflect']:
             self.default_mode = 'loop'
-            warnings.warn("unrecognized default_mode: using 'loop'")
+            _log.warning("unrecognized default_mode: using 'loop'")
```
Since this comes from the caller, I'm not sure it should be logged instead of a warning.

efiring replied:
Attn: @dopplershift, what do you think: should any of the warnings in animation be converted to _log.warning?

dopplershift replied:
It's not clear to me what difference the change makes. I guess warnings from `warnings.warn` don't repeat? In that case I think using the log might be better in that regard.

Something else I'm missing?

jklymak replied:
@dopplershift  `warnings.warns` can't be fine-tuned by the user to appear or not appear.  So, its pretty easy to suppress all matplotlib-specific logging.warning() messages.  I chose to use logging here just because it seemed like a non-critical warning some folks would want to suppress.  OTOH, I'm not sure why I picked on this example.  Happy to do either way..

QuLogic replied:
I think the default logging config does not print a source file or line number.

This warning comes from an invalid input from the user. Actually, I'm not sure why it doesn't just raise an error.

jklymak replied:
Just to re-iterate, I'm not trying to make decisions for the original authors of the modules.  All I did was rip out `verbose` and replace with `logging`.  If the actual contents of the messages need to be changed, or the level/mechanism is wrong, then we can do that as needed.

dopplershift replied:
I think I'm fine with the changes.

efiring replied:
I have no objection to the changes, either.
One additional point of information: warnings can be turned into errors, which can be very useful when debugging.

QuLogic commented on lib/matplotlib/animation.py:
```diff
@@ -969,7 +960,7 @@ def grab_frame(self, **savefig_kwargs):
             self._total_bytes += len(imgdata64)
             if self._total_bytes >= self._bytes_limit:
-                warnings.warn("Animation size has reached {0._total_bytes} "
+                _log.warning("Animation size has reached {0._total_bytes} "
```
Same here.

anntzer commented on lib/matplotlib/figure.py:
```diff
@@ -48,6 +49,8 @@
                                    TransformedBbox)
 from matplotlib.backend_bases import NonGuiException
 
+_log = logging.getLogger(__name__)
```
can do without? (and without the import)

jklymak replied:
I stuck it in there because I have a PR that will need it anyways ;-)  I actually think this should be at the top of most modules so folks use the debugging.  I appreciate there may be a small performance hit on some intensive modules, but hopefully not a big deal

QuLogic commented on lib/matplotlib/backends/backend_gtk.py:
```diff
@@ -3,7 +3,11 @@
 
-import os, sys, warnings
+import logging
+
```
Extra blank line.

QuLogic commented on lib/matplotlib/backends/backend_ps.py:
```diff
@@ -1429,9 +1431,9 @@ def convert_psfrags(tmpfile, psfrags, font_preamble, custom_preamble,
-                               "not enabled the matplotlib 'text.latex.unicode' "
-                               "rcParam.", 'helpful')
+                _log.info("You are using unicode and latex, but have "
+                         "not enabled the matplotlib 'text.latex.unicode' "
```
One more space.

QuLogic commented on lib/matplotlib/dviread.py:
```diff
@@ -21,6 +21,7 @@
                         unicode_literals)
 
 import six
+import logging
```
Should be in the next block.

QuLogic commented on lib/matplotlib/finance.py:
```diff
@@ -17,17 +17,20 @@
 from six.moves.urllib.request import urlopen
 
 import datetime
+import logging
```
As this file is deprecated, you may want to send these changes to the new upstream.

jklymak replied:
Not sure what you mean?  Whatever is replacing this file?  Not sure what that is!

jklymak replied:
Oh `mpl_finance`?  Hmmm

anntzer commented on doc/faq/troubleshooting_faq.rst:
```diff
@@ -114,11 +114,35 @@ provide the following information in your e-mail to the `mailing list
+  command-line version of this, try::
 
-     python simple_plot.py --verbose-helpful > output.txt
+     python -c "from logging import *; basicConfig(level=DEBUG); from pylab import *; plot(); show()"
```
As much as I agree, I think this is better for this specific purpose of getting a log trace.

jklymak commented on doc/devel/contributing.rst:
```diff
@@ -437,6 +437,71 @@ forced to use ``**kwargs``.  An example is
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are five levels at which you can emit messages.
+`logging.critical` and `logging.error`
```
I'll leave as is for now, but am happy for tweaks later...

jklymak commented on lib/matplotlib/__init__.py:
```diff
@@ -215,7 +222,76 @@ def _is_writable_dir(p):
+        oldlev = _log.getEffectiveLevel()
+        if newlev < oldlev:
+            _log.setLevel(newlev)
+            std = {
```
Because if someone has set the level inside their downstream code I want that to win over the deprecated command line.

QuLogic replied:
But what does that have to do with changing where output goes? Is the default always so high that there's no point to changing output locations?

jklymak replied:
Because they (presumably) have set a logging.Handler as well (or are already happy with the default)

QuLogic replied:
Actually, I'm a bit more confused because I can't find any callers of `_set_logger_verbose_level` that set `file_str` to anything but the default.

jklymak replied:
This PR is removing support for `verbose.fileio` matplotlibrc entry, so there should not be a caller.   OTOH, we are still allowing the `--verbose-LEVEL` command line arguments.   

This code could probably go away, but I didn't remove in case someone felt strongly that the matplotlibrc should come back...

jklymak commented on lib/matplotlib/__init__.py:
```diff
@@ -311,7 +387,31 @@ def ge(self, level):
+                wrapper._spoke = spoke
+        return ret
+    wrapper._spoke = False
+    wrapper.__doc__ = func.__doc__
```
To be honest, I just hacked this following what was in the old verbose.wrap.  Not sure what `functools.wraps` is.

jklymak commented on lib/matplotlib/animation.py:
```diff
@@ -569,26 +563,22 @@ def finish(self):
-                               level='helpful')
-                verbose.report("MovieWriter.finish: stderr: %s" % stderr,
-                               level='helpful')
+                _log.info("MovieWriter.finish: stdout: %s", stdout)
```
Not following.  'helpful' was only written if you specified `--verbose-helpful` I think.

QuLogic replied:
Because of [this](https://github.com/matplotlib/matplotlib/pull/9313/files#diff-6814bf77b0702c7e736fb717724a08c5L361).

jklymak commented on doc/devel/contributing.rst:
```diff
@@ -437,6 +437,70 @@ forced to use ``**kwargs``.  An example is
+Using logging for debug messages
+--------------------------------
+
+Matplotlib uses the standard python `logging` library to write verbose
```
@anntzer The doc build gives 

```
Warning, treated as error:
/home/circleci/project/doc/faq/troubleshooting_faq.rst:117:Unknown target name: "logging".
Building HTML failed.
Exited with code 1
```

Did I do this wrong or is our map not working?

jklymak commented:
master rebase

tacaswell commented:
Sweet! :tada:
